Update references to time values to use chrono

harmut01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

References to time should use chrono values. Refactor tests in
connectivity and drivers to remove deprecation warnings due to not
using chrono values.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
None
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
None
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

harrisonmutai-arm committed:
Update references to time values to use chrono

References to time should use chrono values. Refactor tests in
connectivity and drivers to remove depcrecation warninings due to not
using chrono values.

adbridge commented on drivers/tests/TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -73,14 +73,14 @@ void increment_multi_counter(void)
 {
     LowPowerTicker ticker[TICKER_COUNT];
-    const uint32_t extra_wait = 10; // extra 10ms wait time
+    std::chrono::milliseconds extra_wait = 10ms; // extra 10ms wait time
```
rather than poluting the code with std::chrono::milliseconds  in this way you should look at the definition of 'duration' in the events code and do something similar

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -24,10 +24,11 @@
+using namespace std::chrono;
 
-#define ONE_MILLI_SEC 1000
+#define ONE_MILLI_SEC 1ms
```
Pointless #define

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -24,10 +24,11 @@
+#define ONE_MILLI_SEC 1ms
 #define TICKER_COUNT 16
-#define MULTI_TICKER_TIME_MS 100
+#define MULTI_TICKER_TIME_MS 100ms
```
pointless #define

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -90,18 +91,18 @@ void test_case_1x_ticker()
         expected_key = strcmp(_key, "base_time");
     } while (expected_key);
-    greentea_send_kv(_key, callback_trigger_count * ONE_MILLI_SEC);
+    greentea_send_kv(_key, callback_trigger_count * ONE_MILLI_SEC.count());
```
Yuk

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -128,22 +129,22 @@ void test_case_2x_ticker()
     callback_trigger_count = 0;
 
-    ticker1.attach_us(ticker_callback_1, 2 * ONE_MILLI_SEC);
+    ticker1.attach(ticker_callback_1, 2 * ONE_MILLI_SEC);
```
2ms

adbridge reviewed (dismissed):
Need to revisit this, this is not the way to do it IMO, please refer to the events code for examples

harrisonmutai-arm committed:
Refactor updates of time values to chrono

kjbracey commented on drivers/tests/TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -25,11 +25,14 @@
 #define TICKER_COUNT 16
-#define MULTI_TICKER_TIME_MS 100
+#define MULTI_TICKER_TIME_MS 100ms
+#define MULTI_TICKER_TIME_US microseconds{MULTI_TICKER_TIME_MS}
```
This extra define really shouldn't be needed. You can just pass `100ms` to the APIs that take microseconds.

kjbracey commented on drivers/tests/TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -125,11 +128,11 @@ void test_multi_call_time(void)
+        time_diff = gtimer.elapsed_time().count();
 
-        TEST_ASSERT_UINT32_WITHIN(TOLERANCE_US(MULTI_TICKER_TIME_MS * 1000), MULTI_TICKER_TIME_MS * 1000, time_diff);
+        TEST_ASSERT_UINT32_WITHIN(TOLERANCE_US(MULTI_TICKER_TIME_US.count()), MULTI_TICKER_TIME_US.count(), time_diff);
```
I copy-and-pasted a `TEST_ASSERT_DURATION_WITHIN` into a bunch of tests that is a bit neater than doing this. See if you can find it. Didn't manage to spot a nice place to put it for a central definition though. Sorry!

kjbracey commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -162,14 +156,14 @@ void test_case_2x_ticker()
     }
 
-    ThisThread::sleep_for(MULTI_TICKER_TIME_MS + extra_wait);
+    ThisThread::sleep_for(MULTI_TICKER_TIME.count() + extra_wait.count());
```
This is unnecessary, and broken - you're adding a microsecond count to a millisecond count and then passing it to a deprecated non-Chrono API.

I guess the issue was that you were getting a compile error because you were passing microseconds to a millisecond API?

That could have been fixed by

     ThisThread::sleep_for(duration_cast<milliseconds>(MULTI_TICKER_TIME + extra_wait));

but it's actually easier to just not define MULTI_TICKER_TIME as microseconds in the first place. Make it `100ms`, then

    ThisThread::sleep_for(MULTI_TICKER_TIME + extra_wait);

harmut01 replied:
I defined MULTI_TICKER_TIME in microseconds because of TickerBase::attach, however, defining it in terms of ms doesn't cause build errors like sleep_for. Is this because attach is overloaded by some distant base class which accepts milliseconds?

kjbracey replied:
The Chrono system does automatic conversions between resolutions where it's safe.

So if a function takes `microseconds`, and you pass it `milliseconds`, then it does the conversion, via a runtime or compile-time `*1000` on the count.

If the function takes `milliseconds` and you pass it `microseconds`, then you get a compile error due to the loss of precision.  To forcibly lose precision you have to explicitly do `duration_cast<milliseconds>(my_usec_thing)`, and that will then put in a `/1000`.

And this extends to conversions during the arithmetic operations: `5s + 1ms` = `5001ms`.

Basically if it compiles, and it's all Chrono, it does the right thing. The place you can make mistakes is when you use `count()` or constructors to move between Chrono and non-Chrono code, and you get the precision wrong.

Tutorial needed? https://www.youtube.com/watch?v=P32hvk8b13M

kjbracey replied:
Note that there's no leeway for "round constants" on the permitted conversions. It works purely on type, not value. You can't implicitly convert any microsecond value to milliseconds, not even a constant `1000us`.

If you do want a function that can accept arbitrary precision, then you can define it as a template that accepts any sort of `duration`, and you can choose how to do any rounding in the implementation. Functions in the standard C++ library work like this, eg https://en.cppreference.com/w/cpp/thread/sleep_for

But I chose not to do that in our Chrono conversion for simplicity.

harrisonmutai-arm committed:
Add test assertion for chrono values

Remove microsecond macro for ticker time since ThisThread::sleep_for and
TickerBase::attach accept any type of chrono value. Add macro for chrono
value test assertion with and without tolerance.

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -24,17 +24,31 @@
 #define TICKER_COUNT 16
-#define MULTI_TICKER_TIME_MS 100
 volatile uint32_t callback_trigger_count = 0;
 static const int test_timeout = 240;
```
Is test_timeout still used anywhere? If it is should it be a Chrono type too ?

harmut01 replied:
I don't think it should be a chrono type in this case, it's used in GREENTEA_SETUP which only accepts integers.

adbridge commented on drivers/tests/TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -24,21 +24,33 @@
-
+#define TOLERANCE_US(DELAY) (600us + DELAY / 20)
 
+static const int test_timeout = 10;
```
Is test_timeout still used anywhere? If it is should it be a Chrono type too ?

harmut01 replied:
I don't think it should be a chrono type in this case, it's used in GREENTEA_SETUP which only accepts integers.

adbridge commented on drivers/tests/TESTS/mbed_drivers/lp_ticker/main.cpp:
```diff
@@ -143,8 +154,9 @@ void test_detach(void)
     bool ret;
-    const float ticker_time_s = 0.1f;
-    const uint32_t wait_time_ms = 500;
+    const milliseconds ticker_time_s = 10ms;
```
Is this correct? the previous value was 1/10th of a second by the looks of it, the new value is 1/100 th of a second ? Should it not be 100ms ?

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -232,18 +245,18 @@ void test_detach(void)
     bool ret;
-    const float ticker_time_s = 0.1f;
-    const uint32_t wait_time_ms = 500;
+    milliseconds ticker_time = 1000ms;
```
This also looks wrong, you've gone from 1/10th s to 1 s ?

adbridge commented on drivers/tests/TESTS/mbed_drivers/ticker/main.cpp:
```diff
@@ -258,15 +271,16 @@ void test_attach_time(void)
     gtimer.reset();
     gtimer.start();
-    ticker.attach(callback(stop_gtimer_set_flag), ((float)DELAY_US) / 1000000.0f);
+    ticker.attach(callback(stop_gtimer_set_flag), delay);
```
Is this correct ? previous value was microseconds divided by 1000000.0 which is presumably converting to seconds. Your delay conversion above is just doing a direct conversion to milliseconds . Is ticker.attach() thus expecting seconds or milliseconds ? Or are there overloading options to cope with both?

kjbracey replied:
Chrono types perform resolution conversion - implicitly as long as it's lossless, or explicitly via `duration_cast`. You can't mismatch the periods (except by using the `count()` methods or integer constructors so you're going via integers and mismatching there).

In this case it's passing milliseconds to a microseconds function, so there's an implicit *1000 operation happening.

harrisonmutai-arm committed:
Refactor ticker_time_s values

adbridge reviewed (approved):
I believe that looks ok now
